### ISSUE TITLE
Keyboard shortcut documentation for editor in the Wagtail UI

### ISSUE THREAD

**Issue by tmsndrs** [wagtail/wagtail#3949]
Hoping that the proposed rich text editor update will allow better use of keyboard shortcuts (issues documented #2191 ). 

Had a quick think about how we could present the editor with the keyboard shortcuts. 

<img width="812" alt="screen shot 2017-10-19 at 17 24 29" src="https://user-images.githubusercontent.com/7290588/31782427-c51ff1ac-b4f2-11e7-9e2b-a21b06f51f1c.png">

**Comment by tmsndrs** [comment]
Note: I haven't thought about how this modal would be launched yet because I'm unclear if we'll have one set of shortcuts (preferable from a UX perspective I think) or different shortcuts within different parts of the UI.

**Comment by lb-** [comment]
Just adding some basic research into what other popular web applications do in regards activation of keyboard shortcuts modal (via a keyboard shortcut).

It seems that Medium's modal is the only one that can actually be activated within a text box that the user has focus on.

**Medium**
* Open Shortcuts Modal `⌘ + ?` / `Ctrl + ?`
* Close Shortcuts Modal `⌘ + ?` / `Ctrl + ?` not Escape
* Works if you are in the middle of writing text or not.
* Note: Great interface, simple and lets you slide to more advanced shortcuts and small animations on how to do basic things like format text.
* https://help.medium.com/hc/en-us/articles/214672207-Keyboard-shortcuts

**Inbox by Google, Google Drive, Gmail**
* Open Shortcuts Modal `Shift + ?`
* Close Shortcuts Modal `Esc`
* Disabled when inside a text box.
* https://support.google.com/mail/answer/6594?hl=en-GB

**Trello**
* Open Shortcuts Modal `Shift + ?`
* Close Shortcuts Modal `Esc`
* Disabled when writing inside a text box.
* http://help.trello.com/article/798-using-keyboard-shortcuts-in-trello

**Github**
* Open Shortcuts Modal `Shift + ?`
* Close Shortcuts Modal `Esc`
* Disabled when writing inside a text box.
* https://help.github.com/articles/using-keyboard-shortcuts/

**Comment by thibaudcolas** [comment]
Here is the list of keyboard shortcuts supported by Draftail, heavily inspired by Google Docs: https://github.com/springload/draftail/tree/master/docs#keyboard-shortcuts.

I foresee problems integrating this with the rich text editor because the shortcuts would need to change depending on:

- Which editor is used on the current page
- Which rich text features are enabled on the current editor(s)

I imagine we could easily circumvent this though by having some way for rich text editors to hook into the "show keyboard shortcuts" modal, and insert the relevant shortcuts when people are triggering this from a given rich text editor.

**Comment by thibaudcolas** [comment]
It's easier to see how this could work with the improvements done over the last few months:

- The shortcut modal opens with a keyboard shortcut like in examples above
- By default it display's "Wagtail's" shortcuts like in the original screenshot

For rich text shortcuts,

- The shortcut screen hooks into the new "rich text features" API.
- Editors implementing a given rich text feature can define the corresponding shortcut.
- When the shortcut modal opens, if there are rich text features active on the page it grabs these and displays them in a separate section for each active editor.

**Comment by SalahAdDin** [comment]
@thibaudcolas More documentation for `Wagtail 2.0`

**Comment by thibaudcolas** [comment]
@SalahAdDin there's now a list of all Draftail keyboard shortcuts over at https://github.com/springload/draftail/tree/master/docs/user-guide#keyboard-shortcuts (which is referenced in the official Wagtail docs).

**Comment by phildexter** [comment]
@thibaudcolas 

Could we have this on hover?

Miro does this nicely. When you hover over an icon it shows you the shortcut.

![image](https://user-images.githubusercontent.com/34743669/81953890-1d105f00-9600-11ea-892f-66d82076563c.png)

I think we could do a similar thing in Wagtail. When you hover over the Save or Preview buttons we show the shortcut then? This would help with discoverability of these shortcuts too and could be rolled out elsewhere if we brought more in.

**Comment by thibaudcolas** [comment]
I like the idea! Especially considering it’s what Draftail already does, at least for shortcuts that are associated with buttons.

![shortcuts-tooltip](https://user-images.githubusercontent.com/877585/82016908-dcedc280-9679-11ea-84cf-0bd3a411b658.gif)

**Comment by phildexter** [comment]
Lovely, thanks @thibaudcolas !

@benenright what do you think? Are you happy with this approach? If so shall I get some designs done that we can pass by you, or maybe you could do the designs?

**Comment by allcaps** [comment]
This is the way Slack presents keyboard shortcuts. I like it.

<img width="1232" alt="Screenshot 2020-06-06 at 18 46 43" src="https://user-images.githubusercontent.com/1969342/83952050-4eaecb80-a836-11ea-9540-f4cf59684244.png">

Slack has some nice details:
- Main heading in combination with the toggle shortcut. _Press Cmd/ to toggle this panel._
- Shortcuts grouped per section. _Basics, Navigation, et cetera_
- The description is sometimes a link to perform the same action.
- There is an `?` icon for additional information.

Some keyboard shortcuts do _not_ have a visible element:
- Scroll up/down
- Move focus to next/previous section
- Undo/redo
- ...

Wagtail can have those as well. They can _only_ be communicated by a 'cheat sheet' as there is no element to hover.

I think a cheat sheet is the best solution because:
- An overview of all shortcuts is the only way to learn all shortcuts
- A centralised place to display all shortcuts is the best way to hook in custom or third party shortcuts.
- Supports shortcuts that do not have a visible element.

I'd like to help and make the cheat sheet happen. Especially the part where integrators and third parties can hook their own shortcuts to the cheat sheet.

**Comment by phildexter** [comment]
That's great feedback @allcaps. Especially the part about allowing third parties to be able to hook in too. I think a cheat sheet would be useful, but perhaps we could even do both (show some in context where it's useful, but all in the cheat sheet).

The more interesting issue for me is the inconsistency with which the existing shortcuts work foir save and print. Cmd+P often opens the print dialogue, and Cmd+S tries to save the page.

**Comment by allcaps** [comment]
One last thing, shortcuts may be different per operating system.

**Comment by thibaudcolas** [comment]
Interesting approach from Facebook, which has the problem of "shortcuts being different depending on where you are in the app": https://engineering.fb.com/web/facebook-com-accessibility/

**Comment by Stormheg** [comment]
We discussed having keyboard shortcuts in our accessibility team meeting today. There is definitely interest in having this, first step would probably be to create a cheat sheet panel like in Tom's mockup and some way to register them.

There is also the issue that many areas of Wagtail are not currently useable with a keyboard. This is an accessibility issue that we are working on 😅

**Comment by lb-** [comment]
Updated discussion happening in UI chat on slack.

* What if we added a global keyboard shortcut to expand/collapse sidebar - `[` Trello uses this (and uses `]` for the right settings side panel).
* Good feedback on this idea and again general push towards more keyboard shortcuts would be good.
* PR up to do a bit of house cleaning for our current minimal shortcuts approach (note: this is for non-draftail shortcuts) https://github.com/wagtail/wagtail/pull/8900
* @tomdyson  feedback - I increasingly like the ‘command key’ style of keyboard shortcuts, e.g. mod-k for Github and Slack or mod-shift-p in VS Code, where you have one shortcut for bringing up a context-aware modal with type-ahead search on other actions.

![screenshot 2022-07-26 at 14 41 51](https://user-images.githubusercontent.com/1396140/181021929-dacce6bc-f0e4-4c76-b4a5-0eba1a1260df.png)

**Comment by lb-** [comment]
Good article on deciding on keyboard shortcuts 
https://stackoverflow.com/questions/4173707/best-practices-for-designing-keyboard-shortcuts

**Comment by lb-** [comment]
Good reference for common web application keyboard shortcuts https://usethekeyboard.com/

**Comment by lb-** [comment]
See also.  https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/accesskey

Note. It's not recommended to use this but it's a good reference for naming things if we did our own data attributes approach. 

E.g. `data-controller="w-accesskey"`

**Comment by laymonage** [comment]
A suggestion came up in our Wagtail+Next.js workshop:

> ...in the admin editor I love that CTRL+S saves the page. Anyone know the keyboard shortcut for 'Publish' - CTRL+P is the preview

The fact that we don't document the shortcuts (yet people find them anyway) suggests that there's some expectation for this to be useful, so I think this will definitely be a great addition 👍

**Comment by benenright** [comment]
Here are some mockups for how a shortcuts panel could work.

Access from the Help menu (help users to learn the command to toggle the panel to negate the need to access it from this menu):
<img width="815" alt="image" src="https://github.com/wagtail/wagtail/assets/4907819/f4356a6a-c623-4f9e-ad52-11322e2caa9f">

Panel: 
- Scrollable
- Considered making it searchable, but decided to keep things simple for now
- Link out to the full documentation
- Prioritise Wagtail specific commands (currently only 2, bit would be great to add more here eg: Duplicate block, Move block up/down, Add block (outside of richtext), Collapse/Expand block/field, Collapse/Expand all, Toggle comments panel, Toggle status panel, Toggle Minimap ...

<img width="1117" alt="image" src="https://github.com/wagtail/wagtail/assets/4907819/0738df3e-ebfb-4124-9001-525aeba49977">

**Comment by lb-** [comment]
As I know we will want to tackle the keyboard shortcuts stuff better in the coming release or so I have updated the Stimulus adoption issue to be both simpler to use in our DOM and avoids waiting on a Stimulus PR to be merged.

https://github.com/wagtail/wagtail/issues/10167#issuecomment-1949758971

The HTML data attributes to use a keyboard shortcut on any button would be something like; `data-controller="w-kbd" data-w-kbd-key="mod+/"`.

Once that issue has a PR and is merged we can look at a way to collate these controlled elements and their keyboard shortcuts into data and present these in a keyboard shortcut modal. This data would need to incorporate the RichText shortcuts but that should not be too complex.

Still a while off but I feel there is a path forward.

**Comment by thibaudcolas** [comment]
We’ve discussed this with @lb- and @laymonage today, just noting we’re keen to proceed with the above designs with _hard-coded content_ if anyone wants to contribute to this but feels daunted by the scope of having to register shortcuts.

**Comment by lb-** [comment]
Great, I might raise a new issue that covers this smaller scope with a bit of direction, hopefully with enough detail for a new-ish contributor to work with.

**Comment by lb-** [comment]
https://github.com/wagtail/wagtail/issues/11711 has been raised for a very basic initial version of the keyboard shortcuts dialog.

I have also updated this issue's description with a bit of an action plan and links to other issues that have already been created.

I am happy to help anyone that wants to contribute to this with some direction on what to work on or where to help.

**Comment by lb-** [comment]
@benenright - any chance you could add a comment with the keyboard icon svg?

**Comment by lb-** [comment]
We may also want to consider some guidelines in our documentation for how to create accessible keyboard shortcuts. 

See https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-keyshortcuts

Also, the guidelines on that MDN page advise against overriding common browser shortcuts such as `ctrl+p` (for print). Which we do today for the preview tab. 

We probably should be careful to document keyboard shortcuts that may not be good practice.

Thoughts @thibaudcolas ?

**Comment by lb-** [comment]
Also adding this comment from @zerolab 

> I stumbled upon https://github.com/ai/keyux the other day and thought of this issue ([source](https://twitter.com/sitnikcode/status/1760677051127021887))

It's a nice looking library that also has some great documentation on how to make accessible keyboard shortcuts.

**Comment by lb-** [comment]
https://github.com/wagtail/wagtail/pull/11740 has been merged & https://github.com/wagtail/wagtail/issues/10167 has been closed. 

The next steps can be started if anyone is keen, specifically adding a new set of keyboard shortcuts for the sidebar, Minimap and the keyboard shortcuts dialog.

Post here if you want to pick something up and I can create an issue with some clear steps.

**Comment by lb-** [comment]
One new issue created @NXPY123  - slightly different thing but it's something that would be helpful for the usage of these keyboard shortcuts.

https://github.com/wagtail/wagtail/issues/11844

FYI @zerolab  and @allcaps  (I know you Coen raised this concern in one of your other reviews).
